### DISCUSSION TITLE
plotting rhi's created with over the top scanning mode

### DISCUSSION THREAD

**Discussion by srbrodzik** [ARM-DOE/pyart#1166] [Q&A] [answered]
I'm running python3 on a debian 11 system to create plots from SPOL cfradial RHI sector files at various azimuths.  They all plot nicely except for the one scan sequence of 2 rhi's at 101 and 281 deg.  They are created by scanning the radar up at one azimuth and then over the top into the other azimuth.  When I plot the data, only one of the azimuth angles plots correctly.  The other one produces a bunch of blank plots.  I'm thinking it must be the cfradial files themselves.  Would love any advice out there.  Here's the plotting portion of my code.  I don't see a way to submit a sample cfradial file but that would probably be helpful.

`           # create display
            display = pyart.graph.RadarDisplay(radar)
            fig = plt.figure(figsize=(12, 10))
            #fig.suptitle('SPOL '+str(angle)+' deg '+start_str, fontsize=20)
            fig.suptitle('SPOL '+start_str+' UTC RHI '+str(angle)+' deg', fontsize=20)

            ax = fig.add_subplot(321)
            display.plot('DBZ', sweep=index, ax=ax, title='',
                         colorbar_label='Reflectivity (dBZ)',
                         axislabels=('', 'Distance above radar (km)'))
            display.set_limits((0, 150), (0, 18), ax=ax)
    
            ax = fig.add_subplot(322)
            display.plot('VEL', sweep=index, ax=ax,
                         title='', colorbar_label='Radial Velocity (m/s)',
                         cmap='pyart_Carbone42',
                         axislabels=('', ''))
            display.set_limits((0, 150), (0, 18), ax=ax)

            ax = fig.add_subplot(323)
            display.plot('ZDR', sweep=index, ax=ax,vmin=-1,vmax=4,
                         cmap='nipy_spectral',
                         axislabels=('', 'Distance above radar (km)'),
                         title='', colorbar_label='ZDR (dB)')
            display.set_limits((0, 150), (0, 18), ax=ax)
            
            ax = fig.add_subplot(324)
            display.plot('RHOHV', sweep=index, ax=ax,vmin=0,vmax=1,
                         cmap='nipy_spectral',
                         title='', colorbar_label='RHOHV',
                         axislabels=('', ''))
            display.set_limits((0, 150), (0, 18), ax=ax)

            ax = fig.add_subplot(325)
            display.plot('KDP', sweep=index, ax=ax,vmin=-1,vmax=4,
                         cmap='nipy_spectral',
                         title='', colorbar_label='KDP (deg/km)',
                         axislabels=('Distance from radar (km)', 'Distance above radar (km)'))
            display.set_limits((0, 150), (0, 18), ax=ax)

            ax = fig.add_subplot(326)
            display.plot('PHIDP', sweep=index, ax=ax,vmin=-180,vmax=180,
                         cmap='nipy_spectral',
                         title='', colorbar_label='PHIDP (deg)',
                         axislabels=('Distance from radar (km)', ''))
            display.set_limits((0, 150), (0, 18), ax=ax)

            # save image
            #plt.show()
            plt.savefig(outdir+'/'+file_out)

`

**Comment by mgrover1** [comment]
Can you share a sample data file? You can send it to mgrover@anl.gov!

**Comment by mgrover1** [comment]
@srbrodzik - I had a chance to look at this today. Here are some thoughts!

> I'm running python3 on a debian 11 system to create plots from SPOL cfradial RHI sector files at various azimuths. They all plot nicely except for the one scan sequence of 2 rhi's at 101 and 281 deg. 

First, I noticed that we need to manually set the scan mode (assuming these are RHIs), although it sounds like there is a bit of a custom scan strategy here?

```python
radar = pyart.io.read("cfrad.20220531_080000.424_to_20220531_080047.680_SPOL_PrecipRhi1_RHI.nc")
radar.scan_type = 'rhi'
```

I was able to reproduce the issue here, using the data sent over. Here is a plot at each of the azimuths, using this block of code:

```python
cmap_dict = {"DBZ": "pyart_HomeyerRainbow",
             "VEL": "pyart_balance",
             "ZDR": "pyart_Carbone42",
             "RHOHV": "pyart_Carbone42",
             "KDP": "pyart_Carbone42",
             "PHIDP": "pyart_Carbone42",}
display = pyart.graph.RadarDisplay(radar)
for sweep in range(radar.nsweeps):
    for field in cmap_dict.keys():
        display.plot(field, sweep, cmap=cmap_dict[field])
        plt.ylim(0, 20)
        plt.show()
        plt.close()
```

Focusing on just the reflectivity field (`DBZ`), these are the two plots created

![Screen Shot 2022-06-02 at 9 46 18 AM](https://user-images.githubusercontent.com/26660300/171656448-1e7b4f58-578e-44b3-91db-d75374020d24.png)

It looks like it is having an issue with the second one (281 deg)...

> They are created by scanning the radar up at one azimuth and then over the top into the other azimuth. 

Could you expand on this? What do you mean by "over the top into the other azimuth"? Is this a typical RHI scan? A vertically pointing scan?

> I'm thinking it must be the cfradial files themselves. 

I suspect this is the case... maybe it would be helpful to setup a call to chat about this? I know that ya'll are in a very different timezone, so perhaps we could do earlier tomorrow (June 3) early in the morning my time (7:30 AM CDT), which would be 8:30 PM in Taiwan?

**Comment by mgrover1** [comment]
Also, here is a plot of the second RHI zoomed out a bit - there appears to be something picked up by the radar, but I am not exactly sure what it is...
![Screen Shot 2022-06-02 at 10 05 30 AM](https://user-images.githubusercontent.com/26660300/171660751-d5ff9cb1-1c0e-487d-8a26-d6f4d27a0d77.png)

**Comment by srbrodzik** [comment]
Thanks Max.

Let me try your code tomorrow (it's Thursday night around 11pm here right
now) and I'll let you know what I get.  I'm not sure why your code shows
data at 101 deg and mine doesn't.  Would love to talk it over Friday night
my time, Friday morning your time.  We'd have to talk via Signal or
WhatsApp since I currently have a Taiwan phone number. Or we could always
use Zoom.  I've set up Signal and WhatsApp to use my US number so you'd
just have to dial that (206-660-0854) to reach me.  Or I can set up a Zoom
call and share the link with you.

FYI, this is kind of a weird RHI 'volume'.  It goes up at 101 deg (I think
that angle is first but it could be the other way around), stops at the
top, spins the dish around and then comes down at 281 deg.   We're doing
this because there's another radar to the east and this is the line from
one radar to the other.  The second radar will do some similar scans so we
can compare data.

Stacy

**Comment by mgrover1** [comment, marked as answer]
Great! Sounds good -  I will send you a calendar invite for Zoom to the email you used to send the data file. I looped through the different sweeps, so perhaps the sweep number was set to 1 when you tried it? Here is the full script if you want to copy/paste and give it a try. Looking forward to chatting.

```python
import pyart
import matplotlib.pyplot as plt

radar = pyart.io.read("cfrad.20220531_080000.424_to_20220531_080047.680_SPOL_PrecipRhi1_RHI.nc")
radar.scan_type = 'rhi'

cmap_dict = {"DBZ": "pyart_HomeyerRainbow",
             "VEL": "pyart_balance",
             "ZDR": "pyart_Carbone42",
             "RHOHV": "pyart_Carbone42",
             "KDP": "pyart_Carbone42",
             "PHIDP": "pyart_Carbone42",}
display = pyart.graph.RadarDisplay(radar)
for sweep in range(radar.nsweeps):
    for field in cmap_dict.keys():
        display.plot(field, sweep, cmap=cmap_dict[field])
        plt.ylim(0, 20)
        plt.show()
        plt.close()
```

**Comment by srbrodzik** [comment]
Thanks for the help.  I'll check the sweep index in my code when I get to
the Ops Center.  I hope it's not something that stupid:-)

**Comment by srbrodzik** [comment]
Hi Max,

OK, your snippet of code worked for me so I added things from my code to it
until I discovered the problem.  I was putting in xlim = (0,150) but for
this particular angle in this particular scanning sequence, I needed to set
xlim = (-150,0).

Guess we don't need to chat about this tonight.  Thanks for your
willingness to help me figure this out.  I really appreciate it.

I'm sure I'll be in touch down the road.

Stacy

**Comment by srbrodzik** [comment]
I found another data oddness.  If you need me to submit this formally via
github, I'll do that.  If not, here's the story.

We're collecting a different kind of rhi volume that has 12 widely-spaced
azimuths that range from 0 to 360.  They all plot up nicely except the 270
deg azimuth scan which has some artifacts above the radar.  I used your
code snippet to do the plotting and added a couple things so I could tell
one plot from the next.

I thought it might be the cfradial file but the NCAR CIDD software displays
these 270 deg angle rhi's without any artifacts.  See the image called
'spol.202206030411.rhi_270_cidd.png'.

Everything is attached.  Let me know if you have any thoughts.  You'll
probably need to take a look through this so I don't expect to talk about
it tonight.  Just take your time and let me know what you find.  If we need
to chat at that point, we can set something up.

Thanks for your help.  I really appreciate it especially during this field
work.

Stacy

 cfrad.20220603_040943.316_to_20220603_041149.93...
<https://drive.google.com/file/d/1Ij1gccjsSCk6m2ExkdKE5G72KROf0-n5/view?usp=drive_web>

**Comment by mgrover1** [comment]
@srbrodzik coming back to this - it looks like it is a data quality issue.

There are some parts of the 270 degree RHI that have not been flagged to be as "antenna transition" (specifically the last radials)

Using the following manually added in the missing transition flags, but without doing this, there is a "rogue radial" in the top left of the plot causing the issues.

Before fixing the transition data:

![Screen Shot 2022-06-08 at 9 36 37 AM](https://user-images.githubusercontent.com/26660300/172644530-4a9817f8-6518-4ce6-b1e5-814eefbd8e1a.png)

After fixing the transition data:

![Screen Shot 2022-06-08 at 9 38 57 AM](https://user-images.githubusercontent.com/26660300/172644659-04085319-0e2a-46e9-922f-fd51edbcb20e.png)

And here was the code I used to correct it:

```python
import pyart
import matplotlib.pyplot as plt
import numpy as np

# Read in the data and specify rhi scan type
radar = pyart.io.read("cfrad.20220603_040943.316_to_20220603_041149.938_SPOL_PrecipRhi2_RHI.nc")
radar.scan_type = 'rhi'

# Extract the 270 degree rhi which is where the issue is at
subset = radar.extract_sweeps([8])

# Pull out the transition data
trans = subset.antenna_transition["data"]

# Make sure the last four radials are excluded flagged as in transition (1 = in transition)
trans[-4:]  = [1, 1, 1, 1]

# Assign the new values back to the radar
subset.antenna_transition["data"] = trans

# Plot the corrected data
cmap_dict = {"DBZ": "pyart_HomeyerRainbow",
             "VEL": "pyart_balance",
             "ZDR": "pyart_Carbone42",
             "RHOHV": "pyart_Carbone42",
             "KDP": "pyart_Carbone42",
             "PHIDP": "pyart_Carbone42",}

display = pyart.graph.RadarDisplay(subset)
for sweep in range(subset.nsweeps):
    for field in list(cmap_dict.keys()):
        display.plot_rhi(field, sweep, cmap=cmap_dict[field])
        plt.show()
        plt.close()
```

One question for you @srbrodzik - do you mind if we write this up as a blog post? As a debugging case study?

**Reply by srbrodzik** [reply]
Thanks so much for all the help Max.  I'm not sure I would have discovered this 'feature'.

I'll implement your suggestion tomorrow.

**Comment by mgrover1** [comment]
@srbrodzik - can you try updating Py-ART? We cut a new release last week, that should solve this problem.

**Reply by srbrodzik** [reply]
I will try to do that and see what the results are.  Will let you know.

**Reply by srbrodzik** [reply]
Just to verify, is the latest version 1.12.4??

**Comment by mgrover1** [comment]
Yes - the latest version 1.12.4, available on either condaf-forge or pip!

**Reply by srbrodzik** [reply]
I'm having some issues installing the latest version of pyart but am working with our sysadmin to sort them out.

In the meantime, I have other files for which your fix do not work.  I'll email one of them to you so you can make sure your fix is general enough to cover them as well.

**Reply by mgrover1** [reply]
Sounds good!

**Reply by mgrover1** [reply]
I added a more robust solution to this issue in #1185 - we needed more logic with how to calculate the radius from the radar near 270 degrees since some values are negative and some are positive (they should all be positive in this case). Once that is merged and we add a new release, you will be able to use that!

**Reply by srbrodzik** [reply]
Hi again.  I have the newest version of pyart installed (including your recent solution #1185 and I'm still getting bad plots for the 270 deg azimuths.  I will email you my latest attempt.

**Reply by mgrover1** [reply]
@srbrodzik it will not be fixed from installing via pip or condo until we cut a new release (which we will likely do next week). If you are interested in using these fixes before then, please use the following:

```bash
pip install git+https://github.com/ARM-DOE/pyart
```

**Comment by srbrodzik** [comment]
Thanks Max.

**Comment by srbrodzik** [comment]
Since it's already Friday, I think I'll just wait for the new release next
week.  Thanks for all the help.

**Reply by mgrover1** [reply]
@srbrodzik - we released a new version today (1.12.5), which is installable via pip or conda!

**Reply by srbrodzik** [reply]
I'm not sure what's going on but even after upgrading to pyart version 1.12.5 the 270 deg plots are still bad. 

![research Radar_SPOL 20220607233059 rhim_6vars_270](https://user-images.githubusercontent.com/5193622/176560698-08c2b052-8e19-4fba-a4b2-b9c316a46ac2.png)

Should I send you the code and/or cfradial file?

**Reply by kmuehlbauer** [reply]
That's most likely another manifestation of bad transition flags. A robust method for flagging those rays would be needed.

![cfradial_rhi](https://user-images.githubusercontent.com/5821660/176645013-bdade51d-f02a-4b7c-996b-718307e5d9df.png)

**Comment by srbrodzik** [comment]
Thanks Max.

**Comment by srbrodzik** [comment]
When do you plan to update the git?

Stacy

**Reply by mgrover1** [reply]
What do you mean by update the git?

**Comment by mgrover1** [comment]
Yes please! That would be great!

**Comment by srbrodzik** [comment]
Here's the file:
 cfrad.20220607_233059.754_to_20220607_233316.83...
<https://drive.google.com/file/d/18sGJPpjm10UZv0SN1sFF6buSD13uC3QV/view?usp=drive_web>

And the code is attached.  It's meant to take input arguments but I have
some commented out inputs labelled 'FOR TESTING' that you can modify and
use.

**Reply by mgrover1** [reply]
Can you check to see which version of Py-ART is running in your script? I tried reproducing the issue using the data file you sent along, and ended up getting the plot to work correctly.

```python
print(pyart.__version__)
```

Here is the script I used which produced the following plots:

```python
import pyart
import matplotlib.pyplot as plt

radar = pyart.io.read("cfrad.20220607_233059.754_to_20220607_233316.834_SPOL_PrecipRhi2_RHI.nc")

radar.scan_type = "rhi"
cmap_dict = {"DBZ": "pyart_HomeyerRainbow",
             "VEL": "pyart_balance",
             "ZDR": "pyart_Carbone42",
             "RHOHV": "pyart_Carbone42",
             "KDP": "pyart_Carbone42",
             "PHIDP": "pyart_Carbone42",}
display = pyart.graph.RadarDisplay(radar)
for sweep in range(radar.nsweeps):
    print(sweep)
    display.plot("DBZ", sweep, cmap=cmap_dict["DBZ"])
    plt.ylim(0, 15)
    plt.xlim(0 , 100)
    plt.show()
    plt.close()
```

<img width="504" alt="Screen Shot 2022-06-30 at 7 34 53 AM" src="https://user-images.githubusercontent.com/26660300/176678264-740ee674-dbfe-4f2a-8055-df7f30c4977b.png">

**Comment by scollis** [comment]
A general “fix sweep” method would be a nice PR  From: Kai Mühlbauer ***@***.***>Date: Thursday, June 30, 2022 at 4:49 AMTo: ARM-DOE/pyart ***@***.***>Cc: Subscribed ***@***.***>Subject: Re: [ARM-DOE/pyart] plotting rhi's created with over the top scanning mode (Discussion #1166)That's most likely another manifestation of bad transition flags. A robust method for flagging those rays would be needed.—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you are subscribed to this thread.Message ID: ***@***.***>

**Comment by srbrodzik** [comment]
That is odd.  As I told you I thought I had v1.12.5 installed.  I can try
this again but will likely get the same (bad) results.

Can you tell me what routine(s) in pyart you modified to create v 1.12.5?
I can then send you the version of that routine that we have installed and
you can verify (or not) that I have the right version installed.

**Reply by mgrover1** [reply]
We changed the way R (the radius from the radar) is calculated for RHI scans https://github.com/ARM-DOE/pyart/pull/1185/files 

Can you run `conda list` and post the output here?

**Comment by srbrodzik** [comment]
I'm not using conda.  We have a centralized python package manager on our
Debian systems at UW and use either select to install or some version of
apt.  That's why I was asking which routine was modified so that I can look
in /usr/local to see if that's what we have.

I was on the road for a couple days but will try again to run the script
you used to successfully get the 270 plot.  I'll let you know how it goes.

Stacy

**Comment by srbrodzik** [comment]
We uninstalled and reinstalled v 1.12.5.  That did the trick.

I will test this out on a few more volumes and let you know if I have any
other problems.

I really appreciate all the help.

Stacy

**Reply by mgrover1** [reply]
Great! Glad to hear that works. Thanks for working with us + providing feedback on bugs you come across!

**Comment by srbrodzik** [comment]
I just wanted to make sure you had made the new release available on
github.  I've looked there and see that it is.  You can disregard my
question.

Stacy
